### ISSUE TITLE
Softmax by Specific Axis & Output Softmax(QK^T/N) without having to dot with V

### ISSUE THREAD

**Issue by SIV4K** [Dao-AILab/flash-attention#536]
I’m wondering how to apply softmax along a specific axis when using flash attention

Moreover, how Can I obtain the matrix from softmax(QK^T/N) before the dot product with the V matrix and if i could will I lose the benefits of using flash attention? Can speed-up /memory-efficient be apply to just softmax(QK^T/N) 

Thank you

**Comment by tridao** [comment]
Then you should use Pytorch for both. FlashAttention implements attention (softmax(QK^T)V) and doesn't return the intermediate softmax. And softmax is always applied along the last dimension (since that's how attention works).
